### PR TITLE
second major release

### PR DESCRIPTION
**Pull request by haoyang-finix**

### CODE REVIEW

File: finix/exceptions.py [hunk 1]
```diff
@@ -100,7 +101,16 @@ def __init__(self, status=None, reason=None, http_resp=None):
...
+            self.body = http_resp.data.decode('utf-8')
+            self._body_raw = self.body
+            try:
+                tmp_dict = json.loads(self.body)
+                self.body = ErrorList(tmp_dict)
+            except:
+                if self.body == '':
+                    self.body = ErrorList('')
+                else:
+                    self.body = self.body
```
Comment (dominic-pimpinella-finix) [inline finix/exceptions.py:113]: `self.body` for ApiException can be either an ErrorList or a string (if we reach the `else` condition above). May be better to standardize this to always be ErrorList.

File: finix/model/finix_list.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+# finix list implementation
+class FinixList:
+    def __init__(self, some_list, method, **kwargs):
+        self._method = method
+        self._kwargs = kwargs
+
+        if hasattr(some_list, 'embedded'):
+            tmp = some_list.embedded
+            if isinstance(tmp, dict):
```
Comment (dominic-pimpinella-finix) [inline finix/model/finix_list.py:12]: `if isinstance(some_list.embedded)` would work too, not strictly necessary to give some_list.embedded a new name.

File: finix/model/finix_list.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
+from typing import List
+
+
+# finix list implementation
+class FinixList:
+    def __init__(self, some_list, method, **kwargs):
```
Comment (dominic-pimpinella-finix) [inline finix/model/finix_list.py:6]: `some_list` is a bit confusing to me. I see there is `some_list.embedded` and `some_list.links` below, but I'm not 100% sure on what object `some_list` is.

File: finix/model/finix_list.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        return len(self.content)
+    
+    def __getitem__(self,index):
+        return self.content[index]
+
+    def __iter__(self):
+        return iter(self.content)
+    
+    def list_next(self):
+        if self.has_more == False:
```
Comment (dominic-pimpinella-finix) [inline finix/model/finix_list.py:59]: nitpick: `if not self.has_more` is probably more "pythonic"

File: finix/model/finix_list.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        return len(self.errors)
+    
+    def __getitem__(self,index):
+        return self.errors[index]
+
+    def __iter__(self):
+        return iter(self.errors)
+
+    def __str__(self):
+        return "This is an ErrorList object, please customize '__str__' by yourself!"
```
Comment (dominic-pimpinella-finix) [inline finix/model/finix_list.py:170]: nitpick: Do we expect this method to be overridden? If not, why can't we implement `__str__`? Same with FinixList above.

File: finix/model/finix_list.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            self.page = 'no_page_received'
+    
+        self.has_more = False
+        self._next_case = 0
+        if 'next_cursor' in self.page and self.content != []:
+            if self.page['next_cursor'] != 'Null' and self.page['next_cursor'] != 'null':
+                self.has_more = True
+                self._next_case = 1
+        
+        if 'offset' in self.page and self.content != []:
```
Comment (dominic-pimpinella-finix) [inline finix/model/finix_list.py:44]: `offset` and `next_cursor` seem to be mutually exclusive. Is that enforced anywhere? Otherwise, it looks like `offset` will take precedence here.

File: tests/test_list_queries.py [hunk 7]
```diff
@@ -1,227 +1,220 @@
...
-    assert response.page['limit'] == 5
-    assert response.embedded['transfers'][0]['id'][:2] == 'TR'
-
-
-def test_list_reversals(config):
-    client = finix.FinixClient(config)
+from finix.model.finix_list import FinixList, PaginationException
+
+
+def test_list_transfers(client00):
```
Comment (dominic-pimpinella-finix) [inline tests/test_list_queries.py:7]: It's hard to parse what exactly is being tested here, breaking the test up and giving descriptive names like `test_list_next_using_offset`, `test_list_next_using_cursor`, `test_list_next_throws_pagination_error` etc would be helpful.

I think it's important to get these tests right and make sure the FinixList object is well tested.

**Review by dominic-pimpinella-finix — COMMENTED** [review]
Good stuff as always!

Some more thoughts that I didn't mention in-line:

Do we need all the individual commits? If you find them useful, keep them, otherwise we could squash them all into a single commit. In the future, if the commits are distinct, we can have separate, smaller PRs for each.

Do we need to commit spec/finix_openapi_old.yaml?


Nothing major to fix, if you need any clarifications just let me know!

File: finix/api/devices_api.py [hunk 8]
```diff
@@ -26,6 +26,7 @@
 from finix.model.error404_not_found_list import Error404NotFoundList
 from finix.model.error406_not_acceptable import Error406NotAcceptable
 from finix.model.error_generic import ErrorGeneric
+from finix.model.finix_list import FinixList
```
[the remark this answers is not in the document]
Comment (haoyang-finix) [reply finix/api/devices_api.py:29]: _api files are generated from single template, as most api groups has at least one list request, all import FinixList

File: tests/conftest.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
+import pytest
+import finix
+from finix.models import *
+from finix.configuration import Environment, Configuration
+from datetime import datetime
+
+
+@pytest.fixture
+def client00():
```
[the remark this answers is not in the document]
Comment (haoyang-finix) [reply tests/conftest.py:9]: I am not sure about the scope of privileges for those credentials. In the future, we plan to create dedicated credentials for test, and we can clarify by that time

File: finix/model/finix_list.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
+from typing import List
```
[the remark this answers is not in the document]
Comment (haoyang-finix) [reply finix/model/finix_list.py:1]: make sense, gonna make the change

File: finix/exceptions.py (hunk 1 above)
Comment (haoyang-finix) [reply finix/exceptions.py:113]: I want this to always ErrorList so that autocomplete works.

Not sure if autocomplete works if body can be string

File: finix/model/finix_list.py (hunk 3 above)
Comment (haoyang-finix) [reply finix/model/finix_list.py:6]: some_list is the old response of list requests, which has many possible types such as TransferList, AuthorizationList, etc.

File: finix/model/finix_list.py (hunk 5 above)
Comment (haoyang-finix) [reply finix/model/finix_list.py:170]: I did this since I can't think of an elegant yet simple way to print those Lists

I guess users won't directly print a list very often; if they do, I expect them to override this method

File: finix/model/finix_list.py (hunk 6 above)
Comment (haoyang-finix) [reply finix/model/finix_list.py:44]: It's enforced in our APIs. Old APIs use 'offset' and there is no cursor related info, new APIs use 'next_cursor' and there is no 'offset'

File: tests/test_list_queries.py (hunk 7 above)
Comment (haoyang-finix) [reply tests/test_list_queries.py:7]: make a lot of sense, I will look into this

File: tests/conftest.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dominic-pimpinella-finix) [reply tests/conftest.py:9]: I see. Even so, it would be great to have more descriptive names for future developers.

File: finix/model/finix_list.py (hunk 5 above)
Comment (dominic-pimpinella-finix) [reply finix/model/finix_list.py]: I'd just remove __str__ then. We can write a proper __str__ method later if we find we need one.

**Comment by lucas-b-finix** [comment]
@haoyang-finix can you extend the README with how to use lists and how to catch an error?

**Review by dominic-pimpinella-finix — APPROVED** [review]
Thanks for the changes @haoyang-finix, approved and good to merge from my perspective.
